Refactor dynamic dims api, stateless internals, higher level export API

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96699

The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

voznesenskym committed:
[WIP] [RFC] Refactor dynamic dims api, stateless internals, higher level export API

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] [RFC] Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] [RFC] Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] [RFC] Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
A few questions / comments:
1. How does this relate to #96012? Does it subsume it?
2. I see a bunch of `dynamic_dims_range` fields. IMO these should be `dynamic_dims_constraints`. It's OK for the actual set of constrains carried by such fields be just of the MinMaxConstraint kind for now, but hopefully adding to them won't need updating of the corresponding fields; we should be able to just add more kinds of constraint kinds, and associated check functions with such constraint kind.
3. Not in this diff, but I'd like to add another state to the enum, which would mean "mark this dim as dynamic but don't verify, just give me guards you accumulate on it." This is distinct from STATIC; it means I want this dim to be DYNAMIC but for now I don't know what constraints I should put it, I want to "discover" them. I love that we are decoupling verification so I'm hopeful this can be added without a lot of churn over this diff.

voznesenskym commented:
> A few questions / comments:
> 
> 1. How does this relate to [Constrain dynamic api and impl #96012](https://github.com/pytorch/pytorch/pull/96012)? Does it subsume it?

yes 
> 2. I see a bunch of `dynamic_dims_range` fields. IMO these should be `dynamic_dims_constraints`. It's OK for the actual set of constraints carried by such fields be just of the MinMaxConstraint kind for now, but hopefully adding to them won't need updating of the corresponding fields; we should be able to just add more kinds of constraints, and associated check functions with such constraint kinds.

I disagree, it's a range until we say otherwise. I think this might be the last constraint we need for a long time.


> 3. Not in this diff, but I'd like to add another state to the enum, which would mean "mark this dim as dynamic but don't verify, just give me guards you accumulate on it." This is distinct from STATIC; it means I want this dim to be DYNAMIC but for now I don't know what constraints I should put it, I want to "discover" them. I love that we are decoupling verification so I'm hopeful this can be added without a lot of churn over this diff.

I disagree - this structure is purely for allocation. Guard accumulation notions are part of produce_guards. You don't need a new entry in the enum, just change what you pass into produce_guards. Also out of scope for this PR.

voznesenskym committed:
Update on "[WIP] [RFC] Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "[WIP] [RFC] Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1579,6 +1646,16 @@ def _verify(expr, potential_expr):
         if not _simplified:
             for symbol, sources in symbol_to_source.items():
+
+                # TODO(voz): Should this valid range actually come from the passed in dynamic_ranges?
+                # On one hand, yes! It's a lot more like the contract we agreed upon with verification
+                # with regards to produce_guards in the first place.
+                # On the other hand, self.var_to_range is pretty useful to check too!
```
@ezyang opine here please! :)

ezyang replied:
Maybe this is more convenient, but you're mixing up two concerns here.

The user defined range doesn't have to be consistent with the size hint. However, if this happens, it can be tested IMMEDIATELY, whenever you get the user range, you don't have to wait until the end. 

We can also refine the range, e.g., by work done in https://github.com/pytorch/pytorch/pull/96616 But if we refine a range out of what the actual sizevar is, that's an INTERNAL error from the framework.

We want to give different error messages in these two cases. So I prefer not conflating.

voznesenskym commented on torch/_export/__init__.py:
```diff
@@ -203,3 +204,41 @@ def do_not_use_experimental_export(f: Callable, args: Tuple, training=False):
+#         0 <= dynamic_dim(blah, 1) <= 100,
+#     ]
+# )
+def dynamic_dim(t: torch.Tensor, index: int):
+    return (t, (index, None))
```
Note! This PR does *not* implement the range operator API.

ezyang replied:
Can we go straight to having a data structure for this, no tuples please

voznesenskym replied:
Yea

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1253,27 +1278,35 @@ def _get_key(self):
+    def _produce_dyn_sizes(self,
+                           ex: torch.Tensor,
+                           source: Source,
+                           dims: Optional[List[DIM_DYNAMISM_STATE]]) -> List[sympy.Expr]:
```
what does it mean when this is None

voznesenskym replied:
It means we just assume DUCK for all, the back compat state. Hinted at on the `DIM_DYNAMISM_STATE` enum, but probably could be spelled out clearer here. Would you like a comment to that effect?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1253,27 +1278,35 @@ def _get_key(self):
+                                                     ex: torch.Tensor,
+                                                     source: Source,
+                                                     dynamic_dims: Optional[List[DIM_DYNAMISM_STATE]],
+                                                     dynamic_dims_range: Dict[int, "MinMaxConstraint"]):
```
Oh, you implemented up-front? It's weird that dynamic_dims is a list but dynamic_dims_range is a dict; they are both keyed on dims.

voznesenskym replied:
dynamic_dims is 1:1 with the size, every dim has a state.

dynamic_dims_range is only for user marked

ezyang replied:
I mean, I could just as easily say that dynamic_dims is either DYNAMIC or DUCK, and if it's missing then it's STATIC. Would you prefer it if they were all dicts? I mostly just want these to be uniform. (Similarly, Optional[List is a stick in my craw, can you really pass None here?!

voznesenskym replied:
> I mean, I could just as easily say that dynamic_dims is either DYNAMIC or DUCK, and if it's missing then it's STATIC.

True, I do not have a strong preference between that state and this.

> Would you prefer it if they were all dicts? I mostly just want these to be uniform.

If I don't have a strong pref, and you do, yours wins. Let's make them both dicts, and if an index is missing we assume static.

> (Similarly, Optional[List is a stick in my craw, can you really pass None here?!

You asked that above ;) And yes, we can, but we can make this required and remove all "default" notions here...

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1253,27 +1278,35 @@ def _get_key(self):
+            if not is_static:
                 size.append(self.create_symbol(
-                    val, TensorPropertySource(source, TensorProperty.SIZE, i), is_dynamic
+                    val, TensorPropertySource(source, TensorProperty.SIZE, i), dim_state
```
Wouldn't it be more uniform to just have `create_symbol` return an int if dim_state is STATIC? (Rather than testing it here.)

voznesenskym replied:
Yes, but this is the current logic, and I figured out of scope for this change.

ezyang replied:
The smell here is you are doing equality tests against enums. What if you get more enum entries in the future? You have no way of easily finding all sites you need to update (whereas if you have an if-else switch on the enum, it's clear you have to add new handling for it.)

voznesenskym replied:
sg

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1307,6 +1340,16 @@ def create_symbolic_sizes_strides_storage_offset(self, ex: torch.Tensor, source:
         assert all(x is not None for x in stride)
         sym_size = [self.create_symintnode(i, hint=hint) for i, hint in zip(size, ex.size())]
+
+        for i, syms in enumerate(sym_size):
```
nit: syms implies it is plural, but it is not

voznesenskym replied:
Agreed

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1340,8 +1383,9 @@ def create_unbacked_symint(self):
     # but there may be a replacement that allows it to be immediately
     # simplified
-    def create_symbol(self, val: int, source: Source, dyn=False) -> "sympy.Expr":
+    def create_symbol(self, val: int, source: Source, dim_state=DIM_DYNAMISM_STATE) -> "sympy.Expr":
```
The default value here looks wrong

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1403,14 +1443,19 @@ def duck_int(self, val):
     # For convenience in testing, a source is allowed to be a str,
     # in which case we will assume it is a LocalSource
     #
+    # strict_mark_dyn lets you enforce stricter dynamic dim verification. When this flag is set,
```
Do we have any uses cases for strict mark dyn? Export isn't using it (the PoR is that export will allow two dynamic dims to be equal without complaining.)

voznesenskym replied:
strict_mark_dyn is needed for export, I thought? It's not about dim equality, it is about how much constraining you allow. strict = None.

avikchaudhuri replied:
So yeah, `strict_mark_dyn` will cover cases like `s0 < 512` etc.

@ezyang @voznesenskym I'm not entirely sure why we'd make an exception to the rule for equality constraints, though. It sticks out as odd. E.g., we may have discovered that `s1 == s0` (and recorded that in replacements) so we'll end up reporting only that `s0 < 512` whereas `s1 == s0` is a legit constraint that gets "hidden". In my peanut-sized brain I'd imagine we would allow you to specify `s1 == s0` as a user constraint as well (thus expanding our support to more than min/max).

The example I'm thinking about is real here, btw: in https://github.com/pytorch/pytorch/issues/95868 some of the dimensions must be equal (in addition to the range constraint above), and it feels wrong to verify some constraints / report it on one arbitrary dimension that just happens to be the union-find representative of others.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1511,12 +1556,25 @@ def _verify(expr, potential_expr):
             # user directives about relationships, we can remove this check from
             # verification.
             if len(expr.free_symbols) == 1:
-                srcs = symbol_to_source[expr.free_symbols.pop()]
```
This is pre-existing code, but I am confused: why is it OK to pop off a free symbol. This seems bad and potentially violating sympy invariants.

voznesenskym replied:
Interesting, I assumed it was fine, let me double check.

avikchaudhuri replied:
Yeah please remove this. It smells of a hack to maybe prevent duplicate checks?

ezyang replied:
This is still not fixed.

If you have a non-empty set and you want to get the first symbol, `next(iter(...))` works

voznesenskym replied:
Yea, havent gotten to this one yet

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1941,23 +2025,48 @@ def evaluate_expr(self, expr: "sympy.Expr", hint=None):
-    # If we got here, the user did *NOT* mark this dim as dynamic,
-    # but BC behavior is to allocate a symbol anyway.
-    return not assume_static_by_default
+class MinMaxConstraint:
```
Why doesn't this just hold a ValueRange directly

voznesenskym replied:
Hmm, it started off a little different and basically ended up the same. Can refactor.

ezyang replied:
Yes please

ezyang commented on torch/_subclasses/meta_utils.py:
```diff
@@ -160,7 +161,13 @@ def meta_storage(self, s, callback):
+        callback=lambda t: t(),
+        source: Optional[Source] = None,
+        dynamic_dims: Optional[List[DIM_DYNAMISM_STATE]] = None,
+        dynamic_dims_range: Optional[Dict[int, MinMaxConstraint]] = None,
```
Naming this something more generic (e.g., constraints) will save you from having to do a mass rename (and change the type annotation MinMaxConstraint) later, if other constraints show up.

voznesenskym replied:
I dislike generic naming for future refactors, because I do not know when those will show up, if ever, but I do not have strong feelings.

ezyang replied:
My point is that the high level export API is generically named (constraints). If you think that we will never have a constraint besides min-max, then you should lobby harder to make the high level API less generically named too. But I think everyone agrees that it's fine to allow for this possibility. Then I would advocate for consistency between the user facing API and the internal mechanism.

voznesenskym replied:
Sold, can rename.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1253,27 +1278,35 @@ def _get_key(self):
+    def create_symbolic_sizes_strides_storage_offset(self,
+                                                     ex: torch.Tensor,
+                                                     source: Source,
+                                                     dynamic_dims: Optional[List[DIM_DYNAMISM_STATE]],
```
You pass these as kwargs in a bunch of other places, for consistency make these kwargs only too?

voznesenskym replied:
sure thing

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -1379,3 +1404,37 @@ def tensor_always_has_static_shape(
+    An invariant is that the length of this list is the number of dimensions of the tensor.
+    If a dimension is marked in dynamic_ranges, it is set as DYNAMIC.
+    Otherwise, it is set to the default dimension state, either DUCK or STATIC depending on the configuration.
+    """
```
assert that dynamic_shapes config must be True?

voznesenskym replied:
Sure thing

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -507,23 +512,31 @@ def TENSOR_MATCH(self, guard: Guard):
+                if hasattr(value, "_dynamo_dynamic_ranges"):
+                    dims_check = str(dict(value._dynamo_dynamic_ranges))
+                    dims_check = dims_check.replace("-oo", "None")
+                    dims_check = dims_check.replace("oo", "None")
```
This is my skeptical face

voznesenskym replied:
Lol, why?

ezyang replied:
string replacement on string programs is never a good idea

voznesenskym replied:
I agree. Open to feedback on how to do this better.

voznesenskym committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1511,12 +1561,25 @@ def _verify(expr, potential_expr):
                 for src in srcs:
                     if src in dynamic_sources:
-                        raise RuntimeError(f"Attempting to introduce a guard {potential_expr} that violates user's mark_dynamic")
+                        raise RuntimeError(f"Attempting to introduce a guard {potential_expr} that violates user's constraint")
```
So for a symbol that is known to be equal to (replacing) another symbol, both their sources will be in `srcs`, yeah? It will be cleaner / more informative to collect all the sources together and error once reporting all of them. This would indicate not only that, e.g., `s0 < 512` but also `s0` was discovered equal to `s1`. Otherwise what if `s1`'s source is in `dynamic_sources` and not `s0`'s...confusing!

voznesenskym replied:
Sure, we can improve that UX in a followup PR. I think it's kind of a distraction here.

avikchaudhuri replied:
sg

avikchaudhuri commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1530,12 +1593,21 @@ def _verify(expr, potential_expr):
+                        # If this dim is marked dynamic, we need to do a test on it, to ensure that it has not bee
+                        # constrained to an integer.
+                        if _is_int(ss):
+                            raise RuntimeError(f"Attempting to constrain dimension "
```
UX request: It would be sweet if all the verification errors were collected separately and raised together: that way, whether to raise or not could be guarded by the strict flag, but also, the errors could be structured and consumed together. Otherwise users will end up seeing one error at a time, which is not great. Also I want to build some tooling that would consume the errors in their structured form and potentially suggest constraints based off them.

voznesenskym replied:
Feel free to add this mode as a stacked PR on top of this. I do not have strong feelings about the UX.

avikchaudhuri replied:
sg

avikchaudhuri commented on torch/_dynamo/__init__.py:
```diff
@@ -156,14 +164,126 @@ def mark_dynamic(t, index):
+
+
+@forbid_in_graph
+def mark_dynamic_constrain(
```
I really don't like the lock-in to ranges on the one hand (e.g., using names like `dynamic_ranges`) and yet keeping this name generic (instead of, say, `mark_dynamic_ranges`). I do believe we need to add at least equality constraints in the short term, otherwise it doesn't feel complete. But if the argument is that "we'll get there when we get there" and potentially do further refactoring, then why not make it specific here too?

Just trying to understand the logic, I won't die on this hill. :)

voznesenskym committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

avikchaudhuri commented:
Do we really want `clear_dynamic_dims` and `has_dynamic_dims` as part of the public API? Looks like we need docs for those if so, but it feels like they shouldn't be exposed...

voznesenskym commented:
> Do we really want `clear_dynamic_dims` and `has_dynamic_dims` as part of the public API? Looks like we need docs for those if so, but it feels like they shouldn't be exposed...

I kinda like it, because we lack a good mechanism for clearing these. Open to other ideas tho :)

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1885,6 +1967,13 @@ def _simplify_floor_div(self, expr):
+    def _verify_valid_range(self, symbol, valid_range):
+        if symbol not in self.var_to_val:
+            return
+        if has_hint(symbol) and self.size_hint(symbol) not in valid_range:
```
If `symbol in self.var_to_val`, then the symbol MUST have a hint.

ezyang replied:
But I am not sure why you don't just remove the first two lines of this function

voznesenskym replied:
This was raising. I am not sure that invariant works correctly atm?

ezyang replied:
Repro?

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1885,6 +1967,13 @@ def _simplify_floor_div(self, expr):
+            return
+        if has_hint(symbol) and self.size_hint(symbol) not in valid_range:
+            raise RuntimeError(f"Valid range, {valid_range}, contradicts "
+                               f"traced value of {symbol}, {self.size_hint(symbol)}")
```
I want two things here:

1. I want this sanity check at the point where the user said constrain_range / setup the range in the symbol
2. I want a better message, something like "You specified that {source} should fall within {valid_range}, but the concrete value we are tracing with has size {self.size_hint(symbol)} which is inconsistent with the range." Importantly, you must not print the symbol, as this is a meaningless s0; you must print a source of the symbol (or maybe print all of them.)

voznesenskym replied:
> I want this sanity check at the point where the user said constrain_range / setup the range in the symbol

That was there, but then you told me to move it. I can put it back. Which function is the best place for it?

> I want a better message, something like ... 

Sure thing.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1403,14 +1448,19 @@ def duck_int(self, val):
     # intended for normal use
-    def produce_guards(self, placeholders, sources,
-                       source_ref=lambda n: n.name(), *, _simplified=False) -> List[str]:
+    def produce_guards(self, placeholders, sources, dynamic_ranges: Optional[Dict[int, "MinMaxConstraint"]],
```
If you're going to accept None `dynamic_ranges`, might as well make this optional.

ezyang replied:
I think the type signature on dynamic_ranges is wrong. It feels like you are Optional[List[Dict[int, MinMaxConstraint]]

voznesenskym replied:
Yes

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1530,12 +1593,21 @@ def _verify(expr, potential_expr):
-                    dynamic_sources.append(property_source)
+                if dynamic_ranges:
+                    dyn_range = dynamic_ranges[pos]
+                    if dyn_range and i in dyn_range:
```
Why do you need to test `dyn_range`; your type annotation claims that `dyn_range` must be `MinMaxConstraint`

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1530,12 +1593,21 @@ def _verify(expr, potential_expr):
+                        if _is_int(ss):
+                            raise RuntimeError(f"Attempting to constrain dimension "
+                                               f"{source.name()}.size()[{i}] to {int(ss)}, "
+                                               "which violates user's constraints")
```
What user constraint does it violate. The error message will be much better if you can say what constraint is being violated.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1530,12 +1593,21 @@ def _verify(expr, potential_expr):
+                        if vr != _default_value_range(specialize_zero_one=self.specialize_zero_one):
+                            for symbol in ss.node.expr.free_symbols:
+                                self._verify_valid_range(symbol, vr)
+                        dynamic_sources.append(property_source)
```
This still seems weirdly complicated

voznesenskym replied:
What do you prefer?

ezyang replied:
I'm not sure. I haven't understood what the code is doing yet

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1253,27 +1278,37 @@ def _get_key(self):
+                                                     source: Source,
+                                                     *,
+                                                     dynamic_dims: Dict[int, DIM_DYNAMISM_STATE],
+                                                     constraint_dims: Dict[int, "MinMaxConstraint"]):
```
nit: constraint is not an adjective; `constrained_dims`

voznesenskym replied:
https://github.com/pytorch/pytorch/pull/96012#discussion_r1126826538

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -195,6 +196,24 @@ def guard_scalar(a):
 # inclusive both ways
 def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
+    """
+    Constrain range takes a symbol, and records a valid value range for it in the graph.
```
It doesn't take a Symbol; it takes a SymInt or int.

voznesenskym replied:
thanks

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -195,6 +196,24 @@ def guard_scalar(a):
 def constrain_range(a, *, min: Optional[int], max: Optional[int] = None):
+    """
+    Constrain range takes a symbol, and records a valid value range for it in the graph.
+    That value range is the narrowest intersection of the current value range for the symbol.
```
nit: it's weird to say "narrowest" here because the intersection of current and new value range is unique

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -195,6 +196,24 @@ def guard_scalar(a):
+    constrain_range(x, min=3, max=10)
+    constrain_range(x, min=4, max=12)
+
+    Would have a range of (4, 10)
```
nit: conventionally, ( means exclusive, and [ means inclusive. So [4, 10] indicates an inclusive-inclusive range

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -195,6 +196,24 @@ def guard_scalar(a):
+    In this way, it is a one directional api - it can only constrain ranges for a shape further.
+
+    Setting this flag records the range into the user_constrained field on shape_env. See the docs on that field
+    for more information.
```
This last paragraph seems out of date

voznesenskym replied:
It probably is, good catch.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1511,12 +1561,25 @@ def _verify(expr, potential_expr):
+                # We do this by checking of the range is unconstrained*.
+                #
+                # *unconstrained might mean a default unsound 0/1 constraint.
+                if self.var_to_range[symbol] != _default_value_range(specialize_zero_one=self.specialize_zero_one):
```
You asked for me to do it this way. I used to have another way, using a structure.

https://github.com/pytorch/pytorch/pull/96012#discussion_r1128308411

ezyang replied:
I was wrong.

voznesenskym replied:
Rare.

voznesenskym commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1340,8 +1385,12 @@ def create_unbacked_symint(self):
+        if dim_state == DIM_DYNAMISM_STATE.STATIC:
+            return sympy.Integer(val)
+
+        dyn = dim_state == DIM_DYNAMISM_STATE.DYNAMIC
```
That feels like an implementation detail, the contract to take dyn is more consistent

ezyang replied:
> the contract to take dyn is more consistent

But we're entirely talking about internal implementation details here? We just are bikeshedding the name of the local variable `dyn`. Which for the most part is just being used to decide to duck size or not.

voznesenskym replied:
Ah, yeah, misunderstood. Agreed.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -295,6 +314,14 @@ def eval_guards(gm, *args):
 def bind_symbols(gm, *args):
     return gm.shape_env.bind_symbols(fx_placeholder_vals(gm), args)
 
+class DIM_DYNAMISM_STATE(Enum):
```
nit: Conventionally, enum name is camel case, and the enum quantities themselves are all caps.

voznesenskym committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/guards.py:
```diff
@@ -422,9 +422,11 @@ def SHAPE_ENV(self, guard: Guard):
         output_graph = self.check_fn_manager.output_graph
         # NB: self.output_graph can be None in the debug_nops tests
         fs = output_graph.tracked_fakes
+        dyn_ranges_or_indices = [a.dynamic_indices for a in fs]
```
probably want to rename this to `constraints` or something

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -187,19 +187,19 @@ def __init__(
                 allow_scalar_outputs=config.capture_scalar_outputs,
                 allow_dynamic_output_shape_ops=config.capture_dynamic_output_shape_ops,
-                strict_mark_dyn=export,
-                assume_static_by_default=config.assume_static_by_default,
```
If we are still failing we should remove this line (not the one above though)

ezyang commented on torch/_dynamo/utils.py:
```diff
@@ -44,9 +44,16 @@
 from torch.nn.modules.lazy import LazyModuleMixin
 from torch.utils._pytree import tree_flatten, tree_map
 
+DimList = List
```
can remove all the changes from this file

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1059,6 +1089,8 @@ def _clone_input(value):
     fake: Union[FakeTensor, SymInt]
     source: Source
+    # Is None when fake is SymInt
+    dynamic_indices: Optional[Set[int]]
```
I think the type of this isn't right either

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates wether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented:
CI is green omfg

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang committed:
Update on "Refactor dynamic dims api, stateless internals, higher level export API"


The purpose of this API is to execute a few large components of work:

1) Refactor all the internals of plumbing dynamic dimension information after dynamo to be stateless
2) Decouple allocation controls around dynamic dimensions from verification
3) For (2), for allocation, create an enum that dictates whether we are in DUCK (default today), STATIC (aka assume_static_default in the past), or DYNAMIC (aka user constrained, do not duck shape)
4) For (2), for verification, we separate out the list of dynamic ranges entirely from allocation. This means shape_env does not tracking for what we verify on, and instead, it is the callers job to invoke produce_guards() with the various things they want verified, specifically, with the valid ranges. We do use constrain ranges to refine value ranges when doing analysis.
5) We have decided, therefore, as an extension of (4) to double down on "late" checks versus "eager" checks, primarily because the mechanisms for gathering what actually matters happens during guards, and should be a purview of the caller seeking guards, not the shape env. However, for dynamo, these structures are essentially one and the same. 



cc soumith penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge

ezyang commented:
@pytorchbot merge -f "everything is all good"
